### PR TITLE
JSON IO Unit Testing

### PR DESCRIPTION
**Pull request by formalhaunt**
### Description of the Change

Created unit tests for the JSON IO package in the utilities module. Refactored the code to make use of `Optional` where conditional parameters were being returned and provided. Improved comments and simplified some of the code where possible.

Overloaded the load preferences method to allow callers to provide an expected class. This will cause the code to map the loaded JSON into that object rather than returning a JSON Tree. The old methods are still there and this is backwards compatible. I have marked the JSON Tree methods as deprecated. I have also modified the save preferences to take a `Object` rather than a JSON Tree. Now either a JSON Tree or a preference object can be passed and it will be serialized.

The aim of this is a pre-cursor to try and clean up the IO classes that then use those JSON trees. The code is quite complex and results in multiple loops making it very difficult to comprehend what is happening. By serializing and de-serializing the JSON from and into Java POJOs it should become much less complex code. 

### Why Should This Be In Core?

Improves test coverage. Everything is backwards compatible and will help improve code readability in future PRs.

### Benefits

* Improves test coverage
* Simplifies usage of the API

### Possible Drawbacks

N/A

### Verification Process

All unit tests passing and code coverage increased on Sonar.

### CODE REVIEW

File: CoreUtilities/src/au/gov/asd/tac/constellation/utilities/genericjsonio/JsonIO.java [hunk 1]
```diff
@@ -65,230 +76,358 @@ private JsonIO() {
...
+                
+                mapper.writeValue(preferenceFile, rootNode);
+                
+                StatusDisplayer.getDefault().setStatusText(
+                        String.format(
+                                PREFERENCE_FILE_SAVED_MSG_FORMAT,
+                                preferenceFile.getPath()
+                        )
+                );
+            } catch (IOException ex) {
```
Comment (antares1470) [inline CoreUtilities/src/au/gov/asd/tac/constellation/utilities/genericjsonio/JsonIO.java:191]: final

File: CoreUtilities/src/au/gov/asd/tac/constellation/utilities/genericjsonio/JsonIO.java [hunk 2]
```diff
@@ -65,230 +76,358 @@ private JsonIO() {
...
+                final NotifyDescriptor nd = new NotifyDescriptor.Message(
+                        String.format("Can't save preference file: %s", ex.getMessage()),
+                        NotifyDescriptor.ERROR_MESSAGE
+                );
+                DialogDisplayer.getDefault().notify(nd);
```
Comment (antares1470) [inline CoreUtilities/src/au/gov/asd/tac/constellation/utilities/genericjsonio/JsonIO.java:196]: can use `NotifyDisplayer.display()`

File: CoreUtilities/src/au/gov/asd/tac/constellation/utilities/genericjsonio/JsonIO.java [hunk 3]
```diff
@@ -65,230 +76,358 @@ private JsonIO() {
...
+     * @see #loadJsonPreferences(Optional, Optional, Function) 
+     * @deprecated use {@link #loadJsonPreferences(Optional, Optional, Class)} instead
+     */
+    @Deprecated(since = "2.4")
+    public static JsonNode loadJsonPreferences(final Optional<String> loadDir,
+                                               final Optional<String> filePrefix) {
+        return loadJsonPreferences(loadDir, filePrefix, file -> {
+            try {
+                return OBJECT_MAPPER.readTree(file);
+            } catch (IOException ioe) {
```
Comment (antares1470) [inline CoreUtilities/src/au/gov/asd/tac/constellation/utilities/genericjsonio/JsonIO.java:237]: final

File: CoreUtilities/src/au/gov/asd/tac/constellation/utilities/genericjsonio/JsonIO.java [hunk 4]
```diff
@@ -65,230 +76,358 @@ private JsonIO() {
...
+     * @return the de-serialized JSON in the requested format
+     * @see #loadJsonPreferences(Optional, Optional, Function) 
+     */
+    public static <T> T loadJsonPreferences(final Optional<String> loadDir,
+                                            final Optional<String> filePrefix,
+                                            final Class<T> expectedFormat) {
+        return loadJsonPreferences(loadDir, filePrefix, file -> {
+            try {
+                return OBJECT_MAPPER.readValue(file, expectedFormat);
+            } catch (IOException ioe) {
```
Comment (antares1470) [inline CoreUtilities/src/au/gov/asd/tac/constellation/utilities/genericjsonio/JsonIO.java:287]: final

File: CoreUtilities/src/au/gov/asd/tac/constellation/utilities/genericjsonio/JsonIO.java [hunk 5]
```diff
@@ -65,230 +76,358 @@ private JsonIO() {
...
-            } catch (IOException ex) {
-                final String errorMsg = "An error occured reading file ".concat(FilenameEncoder.encode(fileName) + FILE_EXT);
-                LOGGER.log(Level.WARNING, errorMsg, ex);
+                Files.delete(fileToDelete.toPath());
+            } catch (SecurityException | IOException ex) {
+                final NotifyDescriptor nd = new NotifyDescriptor.Message(
+                        String.format("Failed to delete file %s from disk", fileToDelete.getName()),
+                        NotifyDescriptor.ERROR_MESSAGE
+                );
+                DialogDisplayer.getDefault().notify(nd);
```
Comment (antares1470) [inline CoreUtilities/src/au/gov/asd/tac/constellation/utilities/genericjsonio/JsonIO.java:348]: can use `NotifyDisplayer.notify()`

File: CoreUtilities/src/au/gov/asd/tac/constellation/utilities/genericjsonio/JsonIO.java [hunk 6]
```diff
@@ -65,230 +76,358 @@ private JsonIO() {
...
-                // Reconsitute filename to include any file prefix
-                if (!filePrefix.isEmpty()) {
-                    fileName = filePrefix.concat(fileName);
-                }
-                return mapper.readTree(new File(prefDir, FilenameEncoder.encode(fileName) + FILE_EXT));
-            } catch (IOException ex) {
-                final String errorMsg = "An error occured reading file ".concat(FilenameEncoder.encode(fileName) + FILE_EXT);
-                LOGGER.log(Level.WARNING, errorMsg, ex);
+                Files.delete(fileToDelete.toPath());
+            } catch (SecurityException | IOException ex) {
```
Comment (antares1470) [inline CoreUtilities/src/au/gov/asd/tac/constellation/utilities/genericjsonio/JsonIO.java:343]: final

**Review by antares1470 — CHANGES_REQUESTED** [review]

**Comment by formalhaunt** [comment]
Updated code based on comments from @antares1470

**Review by antares1470 — APPROVED** [review]

**Review by arcturus2 — APPROVED** [review]
Thank you very much for this too.

**Pull request merged**